src/braket/default_simulator/simulator.py

```diff
@@ -202,32 +212,96 @@ def _validate_and_consolidate_observable_result_types(
+                DefaultSimulator._assign_observable(
+                    obs_obj,
+                    new_obs,
+                    measured_qubits,
+                    i,
+                    not_none_observable_list,
+                    qubit_observable_mapping,
+                    none_observable_mapping,
+                    identity_qubits,
+                )
+        for i in sorted(identity_qubits):
+            not_none_observable_list.append(Identity([i]))
+        return not_none_observable_list
+
+    @staticmethod
+    def _assign_observable(
+        whole_observable,
+        hashed_qubit_observable,
+        measured_qubits,
+        target_index,
+        not_none_observable_list,
+        qubit_observable_mapping,
+        none_observable_mapping,
+        identity_qubits,
+    ):
+        # Validate that the same observable is requested for a qubit in the result types
+        hashed_qubit_observable = (
+            hashed_qubit_observable[target_index]
+            if isinstance(hashed_qubit_observable, dict)
+            else hashed_qubit_observable
+        )
+        qubit = measured_qubits[target_index]
+        existing_observable = qubit_observable_mapping.get(qubit)
+        if hashed_qubit_observable == Identity.__name__:
+            if qubit not in qubit_observable_mapping:
+                identity_qubits.add(qubit)
+            # Do nothing if non-identity observable already exists on the qubit
+        else:
+            qubit_observable = (
+                whole_observable[target_index]
+                if isinstance(whole_observable, dict)
+                else whole_observable
+            )
+            if not existing_observable:
+                identity_qubits.discard(qubit)
+                qubit_observable_mapping[qubit] = qubit_observable
+
+                if (
+                    not none_observable_mapping.get(hashed_qubit_observable)
+                    and qubit_observable.measured_qubits.index(qubit) == 0
+                ):
+                    not_none_observable_list.append(qubit_observable)
+            else:
+                DefaultSimulator._validate_same_observable(
+                    existing_observable, qubit_observable, qubit
+                )
+
+    @staticmethod
+    def _validate_same_observable(existing, new, qubit):
+        cls_existing = existing.__class__.__name__
+        cls_new = new.__class__.__name__
+        if cls_existing != cls_new:
+            raise ValueError(
+                f"Qubit {qubit} is already being measured in {cls_existing};"
+                f" cannot measure in {cls_new}."
+            )
+        if cls_existing == Hermitian.__name__:
+            if not np.allclose(existing.matrix, new.matrix):
+                raise ValueError(
+                    f"Qubit {qubit} is already being measured in {existing.matrix};"
+                    f" cannot measure in {new.matrix}."
+                )
+            qubits_existing = existing.measured_qubits
```

Comment:
Can you explain what this chunk of code is supposed to do? In particular, are `existing` and `new` supposed to be single-qubit or multi-qubit observables?

Reply:
`existing` and `new` can be either single or multi-qubit _non-tensor product_ obeservables (so the latter case only applies to arbitrary Hermitian observables). The purpose of this method is to ensure that for any qubit that already has an observable measurement, and subsequent non-identity observables are the same (qubit-wise commuting arbitrary Hermitians are not yet supported).

The final branch, which is only reached for multi-qubit observables, ensures that the target qubits don't conflict. For example, let's say we have a 2-qubit Hermitian acting on [1, 2]. If we add the same Hermitian to qubit 1, we want to make sure that 1 is the first target qubit, and the second target is 2, basically ruling out targets of, say, [2, 1] or [0, 1].

Reply:
Thanks for the explanation.